Refactor RequestResponseHandler to a delegation pattern and apply it in Report

mathcoder23 opened the pull request:
# Context:
- Related issue: https://github.com/bisq-network/bisq2/issues/3771 
- This PR refactors `ReportRequestService` and `ReportResponseService` to adopt the `RequestResponseHandler` pattern. The goal is to improve code reusability and centralize the management of P2P connection listening, timeouts, and concurrent request handling, ensuring adherence to the request-response pattern.
- In particular, some services are 'request-only'(leech) and do not adhere to a strict Request/Response model. To accommodate this, this PR creates the LeechRequestResponseHandler for use cases that only require sending requests.

# Changes Made:
- Refactor: ReportResponseService now adopts the RequestResponseHandler pattern (in practice, it only functions in 'response' mode).
- New Abstract Class: Introduced LeechRequestResponseHandler, an abstract class designed for services operating in 'Leech' mode within the P2P network (i.e., they only send requests and do not process responses). ReportRequestService, which forwards requests from a RESTful API to the P2P network, is a typical use case. LeechRequestResponseHandler inherits from RequestResponseHandler but overrides response-handling logic to function solely as a request-side client.
- Refactor: ReportRequestService now adopts the LeechRequestResponseHandler pattern, which eliminates all its manual lifecycle management logic.
- Logging Enhancement: The RequestResponseHandler base class has been enhanced. All its key log outputs now automatically include the specific service class name as a prefix (e.g., [KeepAliveService]), significantly improving debugging efficiency in complex environments.

# Verification & Testing
- I have successfully verified and tested these changes locally.
- Testing was primarily conducted on a CLEAR network using the `seed-node-app` and `node-monitor-web-app`. Since ReportRequestService and ReportResponseService are split into two distinct P2PServiceNode types, "REPORT_RESPONSE" and "REPORT_REQUEST", I have also tested various combinations to verify that both services can operate concurrently on the same Node without interference.

## Test Steps(brief):
1. Configure a CLEAR network for both the seed-node-app and the node-monitor-web-app.
2. Configure different combinations of the "REPORT_RESPONSE" and "REPORT_REQUEST" services in the P2PServiceNode for each app.
3. Access http://localhost:8083/api/v1/report/127.0.0.1:8000. It should return the correct report information.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- Refactor
  - Reworked P2P request/response handling into reusable handler/delegate components, simplifying request routing and response generation.
  - Migrated reporting flows to the new handler pattern and removed legacy per-request helper behavior.
  - Inventory now reads pending-request counts via the unified handler interface.

- Bug Fixes
  - Centralized timeout and disconnect cleanup to reduce stuck or orphaned requests.

- Chores
  - Added a lightweight lifecycle interface to standardize initialize/shutdown across network handlers.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/LeechRequestResponseHandler.java:
```diff
@@ -0,0 +1,35 @@
+import bisq.network.p2p.node.Connection;
+import bisq.network.p2p.node.Node;
+
+public abstract class LeechRequestResponseHandler<T extends Request, R extends Response> extends RequestResponseHandler<T, R> {
```
Maybe a better approach is to split the RequestResponseHandler into the client and server functionality.
Client: send request, listen on response
Server: listen on requests, reply with response

The RequestResponseHandler would then use both (prob. requires delegation as we cannot inherit both, or maybe default functions in the interface are sufficient)

The this class would be just the client and we dont need UnsupportedOperationExceptions.

mathcoder23 replied:
Hi @HenrikJannsen Sounds good. Since interfaces cannot have member variables (such as `requestFuturesByConnectionId`), I chose to refactor the `Request` and `Response` handling using the `delegation pattern.` I reworked the entire Request and Response processing, adding a `BaseHandler` superclass and a `LifecycleHandler` interface to standardize initialization and shutdown management.

`RequestResponseHandler` remains fully compatible with the previous implementation. I also introduced abstract classes `ResponseHandler` and `RequestHandler` for external extension, as well as two basic delegates — `RequestHandlerDelegate` and `ResponseHandlerDelegate` — to handle the core request/response processing.

If you notice any issues, please let me know — I really appreciate your review.

mathcoder23 commented on network/network/src/main/java/bisq/network/p2p/common/RequestHandlerDelegate.java:
```diff
@@ -0,0 +1,84 @@
+                : Optional.empty();
+    }
+
+    protected void processRequest(Connection connection, T request) {
+        log.info("Received {} from {}", StringUtils.truncate(request), connection.getPeerAddress());
+        callback.onRequest(connection, request);
+        R response = callback.createResponse(connection, request);
+        node.sendAsync(response, connection)
+                .whenComplete((result, throwable) -> {
+                    if (throwable != null) {
+                        log.warn("Sending {} to {} failed. {}", StringUtils.truncate(response), connection.getPeerAddress(), ExceptionUtil.getRootCauseMessage(throwable));
+                    } else {
+                        log.info("Sent {} to {}", StringUtils.truncate(response), connection.getPeerAddress());
+                    }
+                });
+    }
```
You are correct. We are not handling exceptions in the Node’s onMessage method. I tested the exception locally, and if we don’t use try/catch, the exception will be swallowed!!!

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/ResponseHandlerDelegate.java:
```diff
@@ -0,0 +1,111 @@
+                return existing;
+            }
+            RequestFuture<T, R> requestFuture = new RequestFuture<>(node, connection, request);
+            requestFuture.orTimeout(timeout, TimeUnit.MILLISECONDS)
+                    .whenComplete((response, throwable) -> {
+                        requestFuturesByConnectionId.remove(k);
+                        if (throwable instanceof TimeoutException) {
+                            log.warn("[{}] Request to {} timed out after {} ms", request.getClass().getSimpleName(), connection.getPeerAddress(), timeout);
+                        } else if (throwable != null) {
+                            log.warn("[{}] Request to {} failed: {}", request.getClass().getSimpleName(), connection.getPeerAddress(), ExceptionUtil.getRootCauseMessage(throwable));
+                        } else {
+                            log.debug("[{}] Request to {} completed", request.getClass().getSimpleName(), connection.getPeerAddress());
+                        }
+                    });
+            return requestFuture;
+        });
```
> I agree with this race condition issue. The requestFuturesByConnectionId.remove(k) comes from an asynchronous callback triggered by the Node listener. Once the compute method returns the requestFuture, the key-level lock is released. If a second thread happens to execute between the moment existing.isDone() == true and the requestFuturesByConnectionId.remove(k) call, the requestFuture created by the second thread can be removed. As a result, requestFuturesByConnectionId.get(connection.getId()) will return null, causing the second thread’s requestFuture to timeout.

I am not that sure about the existing policy to return a pending request in case we would get an additional call on the same connection.
If we would manage the futures not in a map using connectionId but the requestId as key, we could avoid that policy and be more generic. For the use cases in network it mostly does not make sense to send a repeated request before the old one has been received, but it would need some support for order by adding a sequence number to avoid that a response from an older requests overwrites data we got from a more response from more recent state.
With that policy we avoid that so, that simplifies things.... 

Another alternative would be to use computeIfAbsent and return any existing future, also completed (if that tiny race opportunity window).  

Just wanted to throw in those 2 thoughts, not suggesting that we should change it.

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/ResponseHandlerDelegate.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /* --------------------------------------------------------------------- */
+    // LifecycleHandler implementation
```
Small style nit: We usually place the initialize and shutdown after constructor.

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/RequestHandler.java:
```diff
@@ -0,0 +1,94 @@
+
+    @Override
+    public void onMessage(EnvelopePayloadMessage envelopePayloadMessage, Connection connection, NetworkId networkId) {
+        requestHandlerDelegate.resolveRequest(envelopePayloadMessage).ifPresent(request -> processRequest(connection, request));
```
Small style NIT: We usually place the continuation on streams/optionals in the next line of its not very short.
```
requestHandlerDelegate.resolveRequest(envelopePayloadMessage)
    .ifPresent(request -> processRequest(connection, request));
```

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/ResponseHandlerDelegate.java:
```diff
@@ -0,0 +1,111 @@
+        });
+    }
+
+    public Map<String, RequestFuture<T, R>> getRequestFuturesByConnectionId(){
```
We usually use Lombok Getters. But if you want to make it more explicit that its part of the API (what it is), its find to leave it. Motivation for Lombok Getters is primarily to avoid lot of boiler plate by getters/setters but that is here not the case as its only one getter.

HenrikJannsen commented on network/network/src/main/java/bisq/network/p2p/common/HandlerLifecycle.java:
```diff
@@ -17,7 +17,7 @@
 package bisq.network.p2p.common;
 
-public interface LifecycleHandler {
+public interface HandlerLifecycle {
```
Maybe we should move that to common module next to the Service interface. 
I don't have a good name suggestion. In mobile we use `LifecycleAware` but not super happy with that but did not find a better one. If you have some suggestion, please share. 

We could also rename Service to AsyncService and use Service for that one... If we do that lets do it in a dedicated PR with only that refactoring.

HenrikJannsen reviewed (approved):
utACK

Some NITs and discussion points...

mathcoder23 commented:
> utACK
> 
> Some NITs and discussion points...

Hi @HenrikJannsen I’ve addressed all your comments, and renamed the PR title to (Refactor RequestResponseHandler to a delegation pattern and apply it in Report), and performed a rebase.

Compared to your last review, I’ve made the following changes:
- Fix code style use Lombok, line breaks for streams/Optionals, place initialize and shutdown after the constructor
- Add TODOs to refactor Service and Lifecycle naming
- Fix map requestFuturesByConnectionId's race condition
- Add try-catch in Node's onMessage method.
